src/deploy/functions/release/sourceTokenScraper.ts

```diff
@@ -1,28 +1,52 @@
+import { FirebaseError } from "../../../error";
 import { logger } from "../../../logger";
 
+type TokenFetchState = "NONE" | "FETCHING" | "VALID";
+
 /**
  * GCF v1 deploys support reusing a build between function deploys.
  * This class will return a resolved promise for its first call to tokenPromise()
  * and then will always return a promise that is resolved by the poller function.
  */
 export class SourceTokenScraper {
-  private firstCall = true;
-  private resolve!: (token: string) => void;
+  private tokenValidDurationMs;
+  private resolve!: (token?: string) => void;
   private promise: Promise<string | undefined>;
+  private expiry: number | undefined;
+  private fetchState: TokenFetchState;
 
-  constructor() {
+  constructor(validDurationMs = 1500000) {
+    this.tokenValidDurationMs = validDurationMs;
     this.promise = new Promise((resolve) => (this.resolve = resolve));
+    this.fetchState = "NONE";
+  }
+
+  async getToken(): Promise<string | undefined> {
+    if (this.fetchState === "NONE") {
+      this.fetchState = "FETCHING";
+      return undefined;
+    } else if (this.fetchState === "FETCHING") {
+      return this.promise; // wait until we get a source token
+    } else if (this.fetchState === "VALID") {
+      if (this.isTokenExpired()) {
+        this.fetchState = "FETCHING";
+        this.promise = new Promise((resolve) => (this.resolve = resolve));
+        return undefined;
+      }
+      return this.promise;
```

Comment:
Should we throw an error here instead of silently succeeding? Can see argument for either.

Reply:
By throwing an error, do you mean throwing an error and then catching it in an enclosing `try { await getToken(); ... } catch (expiredErr) {...}` block, and then retrying?

Reply:
I mean that in theory we should never reach this codepath - fetchState can only be in one of three states, and we have if condition for all of them. So we could throw (which could happen if we misunderstood this code for somee reason) or just return the promise (which might leave the CLI in an undefined state). I have slight preference for the former.